File: bindings/php/rco.php
<?php

/**
 * Hyper-Scale PHP SDK for RCO Protocol (Ψ-V5.1.0).
 * Optimized for PHP 8.1+ with FFI Preloading support.
 * Sovereign Inversion Technical Reference.
 */
class RcoClient {
    private static $ffi = null;
    private $manifold = null;

    /**
     * Initialize the RCO FFI bridge.
     */
    public static function init(?string $libPath = null) {
        if (self::$ffi !== null) return;

        $libPath = $libPath ?? __DIR__ . '/../../dist/lib/librco_core.so';
        
        if (!file_exists($libPath)) {
            throw new RuntimeException("RCO Sovereign Core not found at: $libPath");
        }

        $cdef = "
            typedef struct rco_manifold_t rco_manifold_t;
            int rco_manifold_init(uint64_t node_id, rco_manifold_t** out_manifold);
            int rco_manifold_project(rco_manifold_t* manifold, const double* rewards, int64_t* results_high, int64_t* results_low, size_t len);
            int rco_manifold_audit(rco_manifold_t* manifold, uint8_t* out_gih);
            bool rco_verify_zk_proof(const uint8_t* proof, size_t len, const uint8_t* public_inputs);
            void rco_manifold_destroy(rco_manifold_t* manifold);
        ";

        try {
            self::$ffi = FFI::cdef($cdef, $libPath);
        } catch (FFI\Exception $e) {
            throw new RuntimeException("RCO Native Library Load Failed: " . $e->getMessage());
        }
    }

    public function __construct(int $nodeId = 1) {
        self::init();
        $handlePtr = FFI::new("rco_manifold_t*");
        $status = self::$ffi->rco_manifold_init($nodeId, FFI::addr($handlePtr));
        if ($status !== 0) {
            throw new RuntimeException("RCO Manifold Initialization Failed: $status");
        }
        $this->manifold = $handlePtr;
    }

    public function project(array $rewards) {
        $count = count($rewards);
        $input = FFI::new("double[$count]");
        $outHigh = FFI::new("int64_t[$count]");
        $outLow = FFI::new("int64_t[$count]");
        
        foreach ($rewards as $i => $r) $input[$i] = (double)$r;
        
        $status = self::$ffi->rco_manifold_project(
            $this->manifold,
            FFI::addr($input), 
            FFI::addr($outHigh), 
            FFI::addr($outLow),
            $count
        );

        if ($status !== 0) throw new RuntimeException("RCO Projection Failed: $status");
        
        $combined = [];
        for ($i = 0; $i < $count; $i++) {
            // Reassemble coordinates (Simplified for PHP 64-bit)
            $combined[] = ["high" => $outHigh[$i], "low" => $outLow[$i]];
        }
        return $combined;
    }

    public function audit(): string {
        $gih = FFI::new("uint8_t[32]");
        $status = self::$ffi->rco_manifold_audit($this->manifold, FFI::addr($gih));
        if ($status !== 0) throw new RuntimeException("RCO Audit Failed: $status");
        return FFI::string($gih, 32);
    }

    public static function verifyZkProof(string $proof, string $publicInputs): bool {
        self::init();
        $proofLen = strlen($proof);
        $p = FFI::new("uint8_t[$proofLen]");
        FFI::memcpy($p, $proof, $proofLen);
        
        $piLen = strlen($publicInputs);
        $pi = FFI::new("uint8_t[$piLen]");
        FFI::memcpy($pi, $publicInputs, $piLen);

        return (bool)self::$ffi->rco_verify_zk_proof($p, $proofLen, $pi);
    }

    public function __destruct() {
        if ($this->manifold) {
            self::$ffi->rco_manifold_destroy($this->manifold);
        }
    }
}

File: bindings/ruby/lib/rco.rb
require 'ffi'

# RCO Ruby SDK (Ψ-V5.1.0)
# Sovereign Inversion Technical Reference

module Rco
  extend FFI::Library
  
  # Load the Sovereign Core library from the dist hierarchy
  LIB_PATH = File.expand_path('../../../dist/lib/librco_core.so', __dir__)
  if File.exist?(LIB_PATH)
    ffi_lib LIB_PATH
  else
    warn "WARNING: RCO Sovereign Core not found at #{LIB_PATH}"
  end

  # Define the Ψ-V5.1.0 ABI signatures
  attach_function :rco_manifold_init, [:uint64, :pointer], :int
  attach_function :rco_manifold_project, [:pointer, :pointer, :pointer, :pointer, :size_t], :int
  attach_function :rco_manifold_audit, [:pointer, :pointer], :int
  attach_function :rco_verify_zk_proof, [:pointer, :size_t, :pointer], :bool
  attach_function :rco_manifold_destroy, [:pointer], :void

  class Manifold
    def initialize(node_id = 1)
      @handle_ptr = FFI::MemoryPointer.new(:pointer)
      status = Rco.rco_manifold_init(node_id, @handle_ptr)
      raise "RCO Manifold Initialization Failed: #{status}" if status != 0
      @handle = @handle_ptr.read_pointer
      
      # RAII: Ensure native resource disposal
      ObjectSpace.define_finalizer(self, self.class.finalize(@handle))
    end

    def self.finalize(handle)
      proc { Rco.rco_manifold_destroy(handle) if handle }
    end

    def project(rewards)
      count = rewards.size
      input = FFI::MemoryPointer.new(:double, count)
      input.put_array_of_double(0, rewards.map(&:to_f))
      
      high_res = FFI::MemoryPointer.new(:int64, count)
      low_res = FFI::MemoryPointer.new(:int64, count)
      
      status = Rco.rco_manifold_project(@handle, input, high_res, low_res, count)
      raise "RCO Projection Failed: #{status}" if status != 0
      
      highs = high_res.get_array_of_int64(0, count)
      lows = low_res.get_array_of_int64(0, count)
      
      # Reassemble into 128-bit coordinates (Ruby handles large ints natively)
      highs.zip(lows).map { |h, l| (h << 64) | (l & 0xFFFFFFFFFFFFFFFF) }
    end

    def audit
      out_ptr = FFI::MemoryPointer.new(:uint8, 32)
      status = Rco.rco_manifold_audit(@handle, out_ptr)
      raise "RCO Audit Failed: #{status}" if status != 0
      out_ptr.read_bytes(32)
    end

    def self.verify_zk_proof(proof, public_inputs)
      p_ptr = FFI::MemoryPointer.new(:uint8, proof.bytesize)
      p_ptr.put_bytes(0, proof)
      
      pi_ptr = FFI::MemoryPointer.new(:uint8, public_inputs.bytesize)
      pi_ptr.put_bytes(0, public_inputs)
      
      Rco.rco_verify_zk_proof(p_ptr, proof.bytesize, pi_ptr)
    end
  end
end
